### PR TITLE
feat(nacos): support grpc java

### LINKED ISSUE #5815: bug: using nacos for service discovery, no support for spring cloud programs?
### Issue description

[error] 47#47: *117481 upstream prematurely closed connection while reading response header from upstream

### Environment

- apisix version (cmd: `apisix version`):
- OS (cmd: `uname -a`):
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
- etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):
- apisix-dashboard version, if have:
- the plugin runner version, if the issue is about a plugin runner (cmd: depended on the kind of runner):
- luarocks version, if the issue is about installation (cmd: `luarocks --version`):

### Steps to reproduce

使用spring cloud grpc程序注册到nacos中，port在meta data中gRPC_port=xxxx

### Actual result

502 Bad Gateway
openresty

### Error log

[error] 47#47: *117481 upstream prematurely closed connection while reading response header from upstream, client: 172.18.0.1

### Expected result

_No response_

### PR DESCRIPTION
**Pull request by vcqr**
### What this PR does / why we need it:

fix #5815 
### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Comment by tzssangglass** [comment]
We need test cases. But previously we found that there were too many nacos instances in the CI and they were taking up too much memory. I wanted to find a way to reduce the number of instances of nacos in the CI.
cc @spacewander @leslie-tsang 

So when reviewing this PR, if the CI passes, we can consider not adding test cases for now.
 what's your optins?

File: apisix/discovery/nacos/init.lua [hunk 1]
```diff
@@ -281,11 +283,18 @@ local function fetch_full_registry(premature)
...
+            local node = {
                 host = host.ip,
                 port = host.port,
                 weight = host.weight or default_weight,
-            })
+            }
+
+            if string.match(scheme, "grpc") and host.metadata.gRPC_port then
+                node.host = host.metadata.gRPC_port
+            end
```
Comment (tzssangglass) [inline apisix/discovery/nacos/init.lua:295]:
```suggestion
            local is_grpc = ngx_re_match(scheme, "grpc", "jo")
            if is_grpc and host.metadata and host.metadata.gRPC_port then
                node.host = host.metadata.gRPC_port
            end
```

and you need to add `local ngx_re_match       = ngx.re.match` at the header of file

File: apisix/discovery/nacos/init.lua [hunk 2]
```diff
@@ -281,11 +283,19 @@ local function fetch_full_registry(premature)
...
+
+            local node = {
                 host = host.ip,
                 port = host.port,
                 weight = host.weight or default_weight,
-            })
+            }
+
+            local is_grpc = ngx_re_match(scheme, "grpc", "jo")
+            if is_grpc and host.metadata and host.metadata.gRPC_port then
```
Comment (spacewander) [inline apisix/discovery/nacos/init.lua:294]: Is there documentation for the gRPC_port? We need to add its link as a comment

File: apisix/discovery/nacos/init.lua [hunk 3]
```diff
@@ -281,11 +283,19 @@ local function fetch_full_registry(premature)
...
-            core.table.insert(nodes, {
+
+            local node = {
                 host = host.ip,
                 port = host.port,
                 weight = host.weight or default_weight,
-            })
+            }
+
+            local is_grpc = ngx_re_match(scheme, "grpc", "jo")
```
Comment (spacewander) [inline apisix/discovery/nacos/init.lua:293]: There are only grpc & grpcs, so `scheme == ...` is enough. No need to use regex.

File: apisix/discovery/nacos/init.lua [hunk 4]
```diff
@@ -281,11 +283,19 @@ local function fetch_full_registry(premature)
...
+            if is_grpc and host.metadata and host.metadata.gRPC_port then
+                node.host = host.metadata.gRPC_port
```
Comment (spacewander) [inline apisix/discovery/nacos/init.lua:295]:
```suggestion
                node.port = host.metadata.gRPC_port
```
Is it a typo?

**Comment by spacewander** [comment]
> We need test cases. But previously we found that there were too many nacos instances in the CI and they were taking up too much memory. I wanted to find a way to reduce the number of instances of nacos in the CI. cc @spacewander @leslie-tsang
> 
> So when reviewing this PR, if the CI passes, we can consider not adding test cases for now. what's your optins?

We can provide mock nacos? Like this one: https://github.com/apache/apisix/blob/3c0c89a58ee872c9564ded3fe6e8998f96235489/t/plugin/aws-lambda.t#L30-L35

**Comment by leslie-tsang** [comment]
> > We need test cases. But previously we found that there were too many nacos instances in the CI and they were taking up too much memory. I wanted to find a way to reduce the number of instances of nacos in the CI. cc @spacewander @leslie-tsang
> > So when reviewing this PR, if the CI passes, we can consider not adding test cases for now. what's your optins?
> 
> We can provide mock nacos? Like this one:
> 
> https://github.com/apache/apisix/blob/3c0c89a58ee872c9564ded3fe6e8998f96235489/t/plugin/aws-lambda.t#L30-L35

Sound good. Or we can add new a PR for test case, merge it until we figure out a new way to reduce the memory cost.

**Comment by tzssangglass** [comment]
> We can provide mock nacos? Like this one:

I'm worried that this isn't real enough, and I'm going to learn about Nacos and modify the way nacos registers services, reducing the instances of Nacos to 1.

**Comment by tzssangglass** [comment]
@vcqr hi, pls fetch upstream master branch to you PR

File: apisix/discovery/nacos/init.lua [hunk 5]
```diff
@@ -281,11 +290,18 @@ local function fetch_full_registry(premature)
...
-            core.table.insert(nodes, {
+
+            local node = {
                 host = host.ip,
                 port = host.port,
                 weight = host.weight or default_weight,
-            })
+            }
+
+            if is_grpc(scheme) and host.metadata and host.metadata.gRPC_port then
```
Comment (spacewander) [inline apisix/discovery/nacos/init.lua:300]: @vcqr 
Before we can merge it, could you add a link to the doc about the usage of gRPC_port?

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**